### PR TITLE
Describe git branching model

### PR DESCRIPTION
**Pull request by eloch216**
This PR attempts to finalize the discussion related to issue #34 by writing down guidelines for using `git-flow` for BioCro development.

### CODE REVIEW

**Review by gsrohde — COMMENTED** [review]
Here are a few preliminary comments.  I'll probably have a few more.

**Comment by justinmcgrath** [comment]
I made some suggestions in a different PR. Although I like the look of a naming scheme like "feature/", "hotfix/", I agree that it's probably going to get confusing since the slash is used to separate remote names from branch names. I think "feature-" etc is a better way to go.

The git-flow site doesn't use "feature-", and I'm not committed to using that, but it seems like it could be helpful at some point.

**Review by gsrohde — COMMENTED** [review]
Ultimately, if the _contributions guidelines_ are intended for general contributors as opposed to project maintainers, this information doesn't really belong here.  All that general contributors really need to know about all of this is that when begining work on a pull request, they should branch off of the `develop` branch.  (Presumably general contributors will not be making hot fixes.)  In fact, perhaps we should add a section called _Making a Pull Request_, and the information about branching off of `develop` could go there.

As I've mentioned before, I think it would make sense to have a maintainer's manual separate from the developer's manual.  Unless and until that is done, I suppose this is as good a place as any for this information.  But in the meantime, I think it would make sense to cordon off sections of this manual of interest only to project maintainers, perhaps with some sort of label at the head of the section, such as **_[For project maintainers:]_**.

One additional thing we should strongly consider: Rather than simply adding a new entry to the NEWS file each time we are preparing a new release on a release branch, we should add a new "[Unreleased]" entry at the top of the NEWS file at the time the release branch is merged into `develop`, after it has been merged into `main`.  (See the section _How can I reduce the effort required to maintain a changelog?_ here: https://keepachangelog.com/en/1.0.0/#effort.)  Both PEcAn and BetyDB do this.  Then, when someone works on a feature on a feature branch, they can fill in this section with the changes made on the branch.  This will lessen the amount of work that needs to be done on a release branch: project maintainers won't need to do a lot of research to figure out everything of significance that has changed since the previous release.  (Of course, if two people are working on features simultaneously, this will increase the likelihood of a merge conflict and may require manually integrating the change messages coming from the two feature branches.)  If we decide to adopt this policy, this also is a bit of information that should be directed at general contributors.

File: developer_documentation/contribution_guidelines.Rmd
Comment (eloch216) [reply developer_documentation/contribution_guidelines.Rmd:123]: > I'm not sure from where you gleaned the convention of using `<branch type>/` prefixes. 
>
> Perhaps you extrapolated from my having named a feature branch `feature/fix-document-quantities-workflow`.

Yes, this is what happened. I am fine with using `-` instead of `/`.

> The Driessen article suggests the branch naming convention `release-1.2` (for example) for a release that is to be tagged `v1.2`, and `hotfix-1.2.1` for a hotfix branch that is to become a release tagged as `v1.2.1`. And for feature branches, he suggests a very loose convention—namely, "anything except `master`, `develop`, `release-*`, or `hotfix-*`".

Okay, we don't need to prefix feature branch names (or "bugfix" branches). However, I don't see a reason to not write out the full version number. I think doing that makes the naming convention simpler.

**Comment by eloch216** [comment]
> Ultimately, if the _contributions guidelines_ are intended for general contributors as opposed to project maintainers, this information doesn't really belong here. All that general contributors really need to know about all of this is that when begining work on a pull request, they should branch off of the `develop` branch. (Presumably general contributors will not be making hot fixes.) In fact, perhaps we should add a section called _Making a Pull Request_, and the information about branching off of `develop` could go there.
> 
> As I've mentioned before, I think it would make sense to have a maintainer's manual separate from the developer's manual. Unless and until that is done, I suppose this is as good a place as any for this information. But in the meantime, I think it would make sense to cordon off sections of this manual of interest only to project maintainers, perhaps with some sort of label at the head of the section, such as **_[For project maintainers:]_**.

I added a new bullet point to hopefully get this point across without needing to make a separate maintainer's manual.

> One additional thing we should strongly consider: Rather than simply adding a new entry to the NEWS file each time we are preparing a new release on a release branch, we should add a new "[Unreleased]" entry at the top of the NEWS file at the time the release branch is merged into `develop`, after it has been merged into `main`.

> If we decide to adopt this policy, this also is a bit of information that should be directed at general contributors.

Good point. I am a little wary of adding this as an additional step in the process of making a release branch that occurs between the two pull requests; I think it will be easy to forget about this. I took a different strategy of recommending that "If there is no `UNRELEASED` section in `NEWS.md` on `develop` when a feature branch is merged into `develop`, one should be added." Let me know if you think this is okay.

File: developer_documentation/contribution_guidelines.Rmd
Comment (eloch216) [reply developer_documentation/contribution_guidelines.Rmd]: I made a change to the text, but not 100% I interpreted this comment the right way.

Comment (justinmcgrath) [reply developer_documentation/contribution_guidelines.Rmd]: When there's a compound modifier, it is can ne hyphenated and sometimes it is necessary to prevent confusion. But if the words aren't being used as a compound modifier, they are never hyphenated. So you could say "an up-to-date changelog" (up-to-date acts as a single adjective) or "the changelog is up to date" (up to date is not an adjective here).

File: developer_documentation/contribution_guidelines.Rmd
Comment (gsrohde) [inline developer_documentation/contribution_guidelines.Rmd:138]: I think this section could be a little clearer, and here is my attempt at redrafting this:

- `NEWS.md` should be updated whenever the package version changes.  Normally, this will happen as part of the process of preparing a new release on a release branch:  A new heading specifying the version number of the new release is placed at the top of the content portion of the _news_ file `NEWS.md`.  This is followed by a listing of the important changes effected by the new release.  To make drafting the contents of this list easier, each time a developer works on a new feature branch, a new draft section labeled `UNRELEASED` should be placed at the top of the _news_ file, with the changes being made by the feature branch being listed underneath it.  Then, when it comes time to merge the feature branch into `develop`, the new `UNRELEASED` section is transferred into the `develop` branch's _news_ file, or, if the `develop` branch already has an `UNRELEASED` section in its _news_ file, the feature branch's `UNRELEASED` section will be integrated into the one on the `develop` branch.  (This process of integrating the two `UNRELEASED` sections will likely be part of resolving an inevitable merge conflict. )

    For details about the format of the _news_ file, see the comment at the top of `NEWS.md`.

    In the case of a hot fix, in addition to whatever substantive changes are made on the hotfix branch, a short section headed by the new release number should be added to the _news_ file to describe the needed fixes the hotfix branch addresses.

(The remark regarding the format of the _news_ file prompts me to suggest that the comment in `NEWS.md` should be revised.  Here's a possible updated version:

> This file should document user-visible changes brought about by each new release.
>
> To facilitate this, when a feature on a feature branch is completed and a pull request is being prepared, a new section should be added at the top of this file under the heading "# Unreleased"; it should list all the important changes made on the feature branch.   All changes should be categorized ...
>
> ... such as "# CHANGES IN BioCro VERSION 2.0.0." This section will combine the draft release notes for all features that have been added since the previous release.

(Omit the last sentence "Subsequent commits will then include a new "Unreleased" section in preparation
for the next release.")

)

**Comment by gsrohde** [comment]
> Good point. I am a little wary of adding this as an additional step in the process of making a release branch that occurs between the two pull requests; I think it will be easy to forget about this. I took a different strategy of recommending that "If there is no `UNRELEASED` section in `NEWS.md` on `develop` when a feature branch is merged into `develop`, one should be added." Let me know if you think this is okay.

In case it isn't clear from my proposed revisions to your revision, I think you're right that it doesn't make sense to add an "[Unreleased]" entry at the top of the NEWS file at the time the release branch is merged into `develop`.  Rather, it makes more sense to add this when one begins (or at least in the course of developing) a new feature branch.

File: developer_documentation/contribution_guidelines.Rmd
Comment (eloch216) [reply developer_documentation/contribution_guidelines.Rmd]: > I think this section could be a little clearer, and here is my attempt at redrafting this:
> 
> The remark regarding the format of the _news_ file prompts me to suggest that the comment in `NEWS.md` should be revised. Here's a possible updated version:

I generally like your suggestions. I tried implementing them with a few modifications.

I think the biggest change was to consolidate all the detailed instructions into the comment in `NEWS.md`, with a short overview in `contribution_guidelines.md`. To me, it seems convenient to have the instructions right there in `NEWS.md` when I start adding new changes.

Let me know what you think.

File: developer_documentation/contribution_guidelines.Rmd
Comment (justinmcgrath) [inline developer_documentation/contribution_guidelines.Rmd:172]: Maybe after "A release branch should be branched off `develop`." add "Update NEWS.md to include an # UNRELEASED section that describes the changes that will be introduced in the feature branch (see NEWS.md for more details)."

File: developer_documentation/contribution_guidelines.Rmd
Comment (gsrohde) [inline developer_documentation/contribution_guidelines.Rmd:143]: Perhaps "of the git-flow branching model"?

**Review by gsrohde — APPROVED** [review]
Generally looks fine to me now.

**Pull request merged**
